### PR TITLE
Feat/tbd 4656 hadoop cluster configuration

### PR DESCRIPTION
**Pull request by bchen-talend**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**
New hadoop configuration retrieve component

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/35/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/35/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/37/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/37/)

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    public HadoopCMClusterService(String serviceName, ServicesResourceV3 cluster, List<String> blacklistParams) {
+        this.serviceName = serviceName;
+        this.cluster = cluster;
+        init(blacklistParams);
+    }
+
+    /**
+     * DOC bchen Comment method "init".
+     */
+    private void init(List<String> blacklistParams) {
```
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java]: Please, slice this method in sub-methods.
Also a code so complex definitely need comments.

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        confs = new HashMap<>();
+        InputStreamDataSource clientConfig = null;
+        try {
+            clientConfig = cluster.getClientConfig(serviceName);
+        } catch (BadRequestException e) {
+            // e.printStackTrace();
+            // ignore the exception, because some service don't contains configuration
```
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java]: Please add a return here.

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        }
+        if (clientConfig == null) {
```
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java]: Can `cluster.getClientConfig(serviceName);` return null?
If not, please remove this part.

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/input/HadoopClusterConfigurationReader.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    public HadoopClusterConfigurationReader(BoundedSource source, HadoopClusterConfigurationInputProperties properties) {
+        super(source);
+        this.properties = properties;
+        initConfigurator();
+    }
+
+    abstract void initBuilder();
+
+    private void initConfigurator() {
```
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/input/HadoopClusterConfigurationReader.java]: PLease, slice this method.

File: components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    static {
+        schema = SchemaBuilder.record("hadoopcluster").fields() //
+                .name("clusterName").type().stringType().noDefault() //
+                .name("serviceName").type().stringType().noDefault() //
+                .name("confFileName").type().stringType().noDefault() //
+                .name("confFileContent").type().stringType().noDefault() //
+                .endRecord();
+    }
+
+    private String[] contents = new String[4];
```
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfiguration.java]: Could you add a comment saying that this weird data structure is forced by an external API?

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopAmbariCluster.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        if (service == HadoopHostedService.HIVE) {
+            ApiConfigFile hcatalogConfig = null;
+            for (ApiConfigFile file : configs) {
+                if ("webhcat-site".equals(file.getType())) { //$NON-NLS-1$
+                    hcatalogConfig = file;
+                    break;
+                }
+            }
+            if (hcatalogConfig != null) {
+                configs.remove(hcatalogConfig);
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopAmbariCluster.java]: can't simple move into loop, as I change the change the list actully

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/v1/RootResourceV1.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * created by bchen on Jun 2, 2015 Detailled comment
+ *
+ */
+import javax.ws.rs.Path;
+
+/**
+ * The root of the Cloudera API. Provides access to all sub-resources available in version 1 of the API.
+ */
+public interface RootResourceV1 {
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/v1/RootResourceV1.java]: yes, there is no v2 yet, but sync with Ambari REST API, Ambari use v1
https://github.com/apache/ambari/tree/trunk/ambari-server/docs/api/v1

File: components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfigurationInputDefinition.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    public Property[] getReturnProperties() {
+        return new Property[0];
+    }
+
+    @Override
+    public RuntimeInfo getRuntimeInfo(ExecutionEngine engine, ComponentProperties properties,
+            ConnectorTopology connectorTopology) {
+        assertEngineCompatibility(engine);
+        //assertConnectorTopologyCompatibility(connectorTopology); TODO not work on Studio
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfigurationInputDefinition.java]: can't reproduce after so many changes on component and studio side then, so uncomment this line

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/AmbariClientBuilder.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+
+/**
+ * created by bchen on Jun 2, 2015 Detailled comment
+ *
+ */
+public class AmbariClientBuilder {
+
+    public static final int DEFAULT_TCP_PORT = 7180;
+
+    public static final long DEFAULT_CONNECTION_TIMEOUT = 0;
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/AmbariClientBuilder.java]: you are right, change to 30s, which is the default time in backend

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiConfigList.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.List;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlElementWrapper;
+import javax.xml.bind.annotation.XmlRootElement;
+
+import com.google.common.base.Objects;
+
+/**
+ * created by bchen on Jun 3, 2015 Detailled comment
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiConfigList.java]: yes, remove all of them, bad comment when I created in eclipse 2 years ago

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiListBase.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import com.google.common.collect.Lists;
+
+/**
+ * created by bchen on Jun 2, 2015 Detailled comment
+ *
+ */
+/**
+ * A generic list.
+ */
+abstract class ApiListBase<T> implements Iterable<T> {
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiListBase.java]: can't recall well, I think there was other attribute exist before, but removed after some changes

File: components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfiguration.java (hunk 5 above)
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-definition/src/main/java/org/talend/components/hadoopcluster/configuration/input/HadoopClusterConfiguration.java]: comment added

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/input/HadoopClusterConfigurationReader.java (hunk 4 above)
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/input/HadoopClusterConfigurationReader.java]: done

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java (hunk 2 above)
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java]: great catch

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java (hunk 1 above)
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopCMClusterService.java]: done

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopAmbariCluster.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        for (ApiActualConfigs actualConfig : actualConfigs) {
+            String serviceName = actualConfig.getServiceInfo().getServiceName();
+            Map<String, String> actualConfigVersions = new HashMap<>();
+            for (ApiComponents component : actualConfig.getComponents()) {
+                for (ApiHostComponents hostComponent : component.getHostComponents()) {
+                    Map<String, Map<String, String>> actualConfigFiles = hostComponent.getHostRoles().getActualConfigs();
+                    for (String configFileType : actualConfigFiles.keySet()) {
+                        String version = ""; //$NON-NLS-1$
+                        // tag for HDP 2.0
+                        version = actualConfigFiles.get(configFileType).get("tag"); //$NON-NLS-1$
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/talend/components/hadoopcluster/runtime/configuration/HadoopAmbariCluster.java]: done

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiConfigFileList2.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+
+import com.google.common.base.Objects;
+
+/**
+ * created by bchen on Jun 3, 2015 Detailled comment
+ *
+ */
+@XmlRootElement(name = "configList")
+public class ApiConfigFileList2 extends ApiListBase<ApiConfigFile> {
```
Comment (bchen-talend) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/model/ApiConfigFileList2.java]: comment added

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/50/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/50/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/51/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/51/)

**Comment by bchen-talend** [comment]
@pbailly sorry, I met problem to quick add runtime test for this component, some comes from the spring dependency problem, and the main is I do not find a good way to mock this service

**Comment by bchen-talend** [comment]
@pbailly anyway add some test for CDH, but ignore before we can set these information on ci-instance

**Comment by bchen-talend** [comment]
waiting for this zendesk https://talend.zendesk.com/hc/en-us/requests/33683

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/59/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/59/)

File: components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/v1/RootResourceV1.java (hunk 7 above)
Comment (pbailly) [inline components/components-hadoopcluster/hadoopcluster-runtime/src/main/java/org/apache/ambari/api/v1/RootResourceV1.java]: Ok, thank you for the explaination

**Comment by pbailly** [comment]
Thank you for the modification, this is way clearer now.
I don't have access to the zendesk, but I trust you to add test when posisble.
Great work!

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/7/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/7/)

**Comment by sgandon** [comment]
@bchen-talend can you check why the build fails here.

**Comment by bchen-talend** [comment]
@sgandon I didn't touch FileRunner class, don't know why it failed, and test it on my local, it works well, so I'm going to rebuild one time

**Comment by gonchar-ivan** [comment]
I met the same fail on my PR

**Comment by bchen-talend** [comment]
Hi there,
To unlock TUP team, I merge this component even the integration test enviroment is not prepared, will active these test after https://jira.talendforge.org/browse/DEVOPS-2202 in later PR

**Pull request merged**